### DISCUSSION TITLE
[FEEDBACK REQUEST] About my listeners

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#47] [General]
https://gist.github.com/busslina/7bfd02ccd1e615220d591d5a6c184d86

Firstly, I have two types of listener, so don't confuse them:
1. Resource listener (websocket connections).
2. Connection state change listeners.

My structure is:
- Connection: Includes two capsules, the `Connection` class and the `ConnectionController` class (plus other helper classes).
- Resource: Mainly composed by the `asResource` Side Effect, which is formed by `_ResourceValueWrapper`, `_ResourceListeners` and `_ResourceLock` classes.

The problematic: (Side Effects cannot use other capsules, so to be able inside `asResource` to listen a specific connection state change it requires a `Connection` getter parameter, `getConnection`). Then, inside `_ResourceListeners` and `_ResourceLock` classes it is used (`getConnection`) to listen to specific connection state changes, using a helper class `ConnectionStateListenerManager` (see `connection.dart`) that allows only one listener per connection.

That's it. I tried to do as simpler and smarter as I could.

You can note that I decided the connection capsules to be immutable. But I don't think that making them reactive would solve/simplify anything here.

Any feedback/correction will be highly appreciated.

**Comment by busslina** [comment]
I just found that I can change this:
```dart
cc.read(connectionCapsule)(listener).sendMessageResource(messageData);
```

with:

```dart
_getConnection(listener).sendMessageResource(messageData);
```

But it doesn't change anything on this issue.

**Comment by busslina** [comment]
Let say that, now, doing the Flutter part, I adopted a better practice: I still have helper classes along with the Side Effect, but not returning them, returning some of their methods instead.

I will readapt it too for Backend, but again, it doesn't change anything on this issue.

**Reply by GregoryConrad** [reply]
> I still have helper classes along with the Side Effect, but not returning them, returning some of their methods instead.

That often is a good solution. I tend to gravitate toward that approach in my own code, but then again, I basically never make a new class unless it is for a new widget (when I am required to).

**Comment by GregoryConrad** [comment]
There's a bit too much code to review here for me, but I'll try to give some feedback on what you said you did.

> The problematic: (Side Effects cannot use other capsules, so to be able inside asResource to listen a specific connection state change it requires a Connection getter parameter, getConnection). Then, inside _ResourceListeners and _ResourceLock classes it is used (getConnection) to listen to specific connection state changes,

That all seems fine.

>  using a helper class ConnectionStateListenerManager (see connection.dart) that allows only one listener per connection.

Not so sure about that and making `FooBarController`s in general when using ReArch. When interfacing with an external controller, you obviously need to work with controllers, but when making your own code, you often can avoid making controllers by refactoring your code in some way. Controllers tend to break the whole reactive paradigm because they like to centralize some logic in an otherwise decentralized way of thinking.
